File: A14_Merge_Sort_Equipo_10/merge_sort.py
def merge_sort(nums):
    # Base case
    if len(nums) <= 1:
        return nums
    
    # Divide

    middle = len(nums) // 2
    left_list = nums[:middle]
    right_list = nums[middle:]

    merge_sort(left_list)
    merge_sort(right_list)

    # Conquer

    # The i index is for the left_list, it is pointing
    # to the current index of the left_list
    i = 0
    # The j index is for the right_list, it is pointing
    # to the current index of the right_list
    j = 0
    # The k index is for the nums list, it is pointing
    # to the current index of the nums list
    k = 0

    #Here we will compare the current index of the left_list
    #with the current index of the right_list (the values on each index)
    while i < len(left_list) and j < len(right_list):
        # If the value on the left_list is lower, we set that value
        # to the nums list
        # and increment the i index
        if left_list[i] < right_list[j]:
            nums[k] = left_list[i]
            i += 1
        else:
            # If the value on the right_list is lower, we set that value
            # to the nums list
            # and increment the j index
            nums[k] = right_list[j]
            j += 1
        
        # Increment the k index because
        # we set a value on the nums list in the if/else statements above
        k += 1
    
    #After we do the while loop above, we will have some values left
    #on either the left_list or the right_list
    #We need to add those values to the nums list
    #This is for the left_list
    while i < len(left_list):
        nums[k] = left_list[i]
        i += 1
        k += 1
    
    #This is for the right_list
    while j < len(right_list):
        nums[k] = right_list[j]
        j += 1
        k += 1

A = [38, 27, 43, 3, 9, 82, 10]
B = [ 5, 2, 7, 6, 2, 1, 0, 3, 9, 4]

merge_sort(A)
merge_sort(B)
print(A)
print(B)

File: A14_Merge_Sort_Equipo_10/merge_sort.cpp
#include <vector>
#include <iostream>

using namespace std;

void merge(vector<int>& arr, int s, int m, int e) {
    // Copy the sorted left & right halfs to temp arrays
    vector<int> L = {arr.begin() + s, arr.begin() + m + 1};
    vector<int> R = {arr.begin() + m + 1, arr.begin() + e + 1}; 

    int i = 0; // index for L
    int j = 0; // index for R
    int k = s; // index for arr

    while (i < L.size() && j < R.size()) {
        if (L[i] <= R[j]) {
            arr[k] = L[i++];
        } else {
            arr[k] = R[j++];
        }
        k++;
    }

    // One of the halfs will have elements remaining
    while (i < L.size()) {
        arr[k++] = L[i++];
    }
    while (j < R.size()) {
        arr[k++] = R[j++];
    }
}

vector<int> mergeSort(vector<int>& arr, int s, int e) {
    if (s >= e) {
        return arr;
    }
    // The middle index of the array
    int m = (s + e)  / 2;

    // Sort the left half
    mergeSort(arr, s, m);

    // Sort the right half
    mergeSort(arr, m + 1, e);

    // Merge sorted halfs
    merge(arr, s, m, e);
    
    return arr;
}

int main() {
    vector<int> arr = {5, 4, 3, 2, 1};
    mergeSort(arr, 0, arr.size() - 1);
    // Print
    for (int i = 0; i < arr.size(); i++) {
        cout << arr[i] << " ";
    }
    return 0;
}